x-pack/platform/plugins/shared/streams/server/lib/rules/esql/types.ts

```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { z } from '@kbn/zod';
+
+export type EsqlRuleInstanceState = z.infer<typeof esqlRuleInstanceState>;
+export const esqlRuleInstanceState = z.object({
+  previousOriginalDocumentIds: z.string().array().optional(),
+});
+
+export type EsqlRuleParams = z.infer<typeof esqlRuleParams>;
+export const esqlRuleParams = z.object({
+  query: z.string(),
+  timestampField: z.string(),
+});
```

Comment:
sorry for the late pass. I don't think inferring types from Zod is a good idea given that it creates types that are hard to read, has an impact on TypeScript responsiveness and is ambiguous in terms of input vs output. Can we use a concrete type and :z.Schema<EsqlRuleInstanceState>?

Reply:
sure

Reply:
I can't cast my schema definition to a `z.Schema<Type>` because the rule registration function doesn't accept a generic Zod schema (otherwise, it would allow something like `z.string()` as a rule params schema, which isn't valid).
As a result, I can remove the use of `z.infer` and manually define the interface, but this means I can't enforce that the schema and the interface always stay in sync.

Reply:
1) Christ... 2) `satisfies`?

Reply:
I've yet to see how I can make a schema satisfies a type or vice-versa (without using z.infer 😂 )

Reply:
```
import { z } from 'zod';

type EsqlRuleInstanceState = {
  previousOriginalDocumentIds?: string[]
};

export const esqlRuleInstanceState = z.object({
  previousOriginalDocumentIds: z.string().array().optional(),
}) satisfies z.Schema<EsqlRuleInstanceState>;
```

https://www.typescriptlang.org/play/?ssl=9&ssc=46&pln=1&pc=1#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgNwBQFMAnmAKZwCiAzgI4A2ASgK4f0BJAHYsYAQyEBjegGVxMRgF5EFOHDBR6AN2AQeLAPJRgAc2BCxHACIRJPEPSEwBxFgH4AXHFHGhJgNoAuhRolBT0AB6QsHCSECLw9OzcfIIJEtJyYgpwysgAdBAARgBW9JIwABQIquqaOnqGxmYW1rb2js6uXgU+5iaVAJT5YlBQYjRDhWAwuq1DADQhg97ZwCyYwEko+TKSABb0IGIAPKycvPzCohmy8vQAfGRAA

Reply:
and now i'm wondering, is that faster than z.infer?

Reply:
It changes nothing for the schema, but it does for the types